File: backend-csharp/Controllers/AuthController.cs
using backend_csharp.Data;
using backend_csharp.Models;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace backend_csharp.Controllers;

[Route("api/v1/auth")]
public class AuthController(ExamDbContext dbContext) : ApiControllerBase
{
    [HttpPost("login")]
    public async Task<IActionResult> Login([FromBody] LoginPayload payload)
    {
        string identifier = (payload.Identifier ?? payload.Email ?? string.Empty).Trim().ToLowerInvariant();
        NguoiDung? user = await dbContext.NguoiDung
            .Include(record => record.VaiTro)
            .Include(record => record.SinhVien)
            .FirstOrDefaultAsync(record => record.Email.ToLower() == identifier);

        if (user is null || !user.TrangThai || !PasswordMatches(user, payload.Password))
        {
            return Unauthorized(Error("Email or password is incorrect."));
        }

        return Ok(SessionPayload(user, "csharp-auth"));
    }

    [HttpPost("signup")]
    public async Task<IActionResult> Signup([FromBody] SignupPayload payload)
    {
        int roleId = payload.VaiTroID <= 0 ? 4 : payload.VaiTroID;
        string code = FirstPresent(payload.MaDinhDanh, payload.MaSinhVien, await NextCode(roleId));
        NguoiDung user = new()
        {
            Email = payload.Email.Trim().ToLowerInvariant(),
            HoTen = string.IsNullOrWhiteSpace(payload.HoTen) ? payload.Email : payload.HoTen.Trim(),
            MatKhauHash = NguoiDungController.NormalizePassword(payload.MatKhau),
            VaiTroID = roleId,
            TrangThai = true,
            TaoLuc = DateTime.Now
        };

        dbContext.NguoiDung.Add(user);
        try
        {
            await dbContext.SaveChangesAsync();
            if (roleId == 4)
            {
                dbContext.SinhVien.Add(new SinhVien
                {
                    NguoiDungID = user.NguoiDungID,
                    MaSinhVien = code,
                    HoTen = user.HoTen,
                    Email = user.Email,
                    TrangThai = true,
                    TaoLuc = DateTime.Now
                });
                await dbContext.SaveChangesAsync();
            }

            await dbContext.Entry(user).Reference(record => record.VaiTro).LoadAsync();
            await dbContext.Entry(user).Reference(record => record.SinhVien).LoadAsync();
            Dictionary<string, object?> response = SessionPayload(user, "csharp-auth-signup");
            response["generatedCode"] = code;
            return Created("/api/v1/auth/signup", response);
        }
        catch (DbUpdateException exception) when (IsDatabaseConflict(exception))
        {
            return Conflict(Error("Email or generated code already exists."));
        }
    }

    private static bool PasswordMatches(NguoiDung user, string password)
    {
        string plain = user.MatKhauHash.StartsWith("hashed_", StringComparison.Ordinal) ? user.MatKhauHash[7..] : user.MatKhauHash;
        return user.MatKhauHash == password || plain == password;
    }

    private async Task<string> NextCode(int roleId)
    {
        string prefix = roleId == 2 ? "CBDT" : roleId == 3 ? "CBKT" : "SV";
        List<string> codes = await dbContext.SinhVien
            .Where(record => record.MaSinhVien.StartsWith(prefix))
            .Select(record => record.MaSinhVien)
            .ToListAsync();

        int nextNumber = codes
            .Select(code => new string(code.Where(char.IsDigit).ToArray()))
            .Select(number => int.TryParse(number, out int parsed) ? parsed : 0)
            .DefaultIfEmpty(0)
            .Max() + 1;

        return $"{prefix}{nextNumber.ToString().PadLeft(3, '0')}";
    }

    private static string FirstPresent(params string?[] values)
    {
        return values.FirstOrDefault(value => !string.IsNullOrWhiteSpace(value))!.Trim();
    }

    private static Dictionary<string, object?> SessionPayload(NguoiDung user, string source) => new()
    {
        ["token"] = $"csharp-{user.NguoiDungID}-{user.VaiTro?.TenVaiTro}",
        ["authSource"] = source,
        ["user"] = NguoiDungController.SerializeUser(user)
    };

    public sealed class LoginPayload
    {
        public string? Identifier { get; set; }
        public string? Email { get; set; }
        public string Password { get; set; } = string.Empty;
    }

    public sealed class SignupPayload
    {
        public string HoTen { get; set; } = string.Empty;
        public string Email { get; set; } = string.Empty;
        public string MatKhau { get; set; } = string.Empty;
        public int VaiTroID { get; set; } = 4;
        public string? MaDinhDanh { get; set; }
        public string? MaSinhVien { get; set; }
    }
}

File: backend-ruby/app/controllers/api/v1/auth_controller.rb
class Api::V1::AuthController < ApplicationController
  def login
    identifier = params[:identifier].presence || params[:email]
    password = params[:password].to_s
    user = NguoiDung.includes(:vai_tro, :sinh_vien).find_by("LOWER(Email) = ?", identifier.to_s.downcase)

    unless user && user.TrangThai && password_matches?(user, password)
      render json: { error: "Email or password is incorrect." }, status: :unauthorized
      return
    end

    render json: session_payload(user, "ruby-auth"), status: :ok
  end

  def signup
    role_id = (params[:VaiTroID] || 4).to_i
    code = params[:MaDinhDanh].presence || params[:MaSinhVien].presence || next_code(role_id)
    user = NguoiDung.new(
      Email: params[:Email].to_s.strip.downcase,
      HoTen: params[:HoTen].presence || params[:Email],
      MatKhauHash: normalize_password(params[:MatKhau]),
      VaiTroID: role_id,
      TrangThai: true
    )

    if user.save
      sync_student_code(user, code)
      render json: session_payload(user.reload, "ruby-auth-signup").merge(generatedCode: code), status: :created
    else
      render json: { errors: user.errors.full_messages }, status: :unprocessable_entity
    end
  rescue ActiveRecord::RecordNotUnique
    render json: { error: "Email or generated code already exists." }, status: :conflict
  end

  private

  def password_matches?(user, password)
    [ user.MatKhauHash, user.MatKhauHash.to_s.sub(/^hashed_/, "") ].include?(password)
  end

  def normalize_password(value)
    text = value.to_s.strip
    text.start_with?("hashed_") ? text : "hashed_#{text}"
  end

  def next_code(role_id)
    prefix = role_id == 2 ? "CBDT" : role_id == 3 ? "CBKT" : "SV"
    existing = SinhVien.where("MaSinhVien LIKE ?", "#{prefix}%").pluck(:MaSinhVien)
    number = existing.map { |code| code.to_s.delete("^0-9").to_i }.max.to_i + 1
    "#{prefix}#{number.to_s.rjust(3, "0")}"
  end

  def sync_student_code(user, code)
    return unless user.VaiTroID.to_i == 4 && code.present?

    student = user.sinh_vien || SinhVien.find_or_initialize_by(MaSinhVien: code)
    student.NguoiDungID = user.NguoiDungID
    student.MaSinhVien = code
    student.HoTen = user.HoTen
    student.Email = user.Email
    student.TrangThai = true
    student.save!
  end

  def session_payload(user, source)
    {
      token: "ruby-#{user.NguoiDungID}-#{user.vai_tro&.TenVaiTro}",
      authSource: source,
      user: {
        NguoiDungID: user.NguoiDungID,
        Email: user.Email,
        MatKhauHash: user.MatKhauHash,
        HoTen: user.HoTen,
        VaiTroID: user.VaiTroID,
        TenVaiTro: user.vai_tro&.TenVaiTro,
        TrangThai: user.TrangThai ? 1 : 0,
        MaSinhVien: user.sinh_vien&.MaSinhVien,
        TaoLuc: user.TaoLuc,
        CapNhatLuc: user.CapNhatLuc
      }
    }
  end
end
